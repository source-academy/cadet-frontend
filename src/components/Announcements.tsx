/**
 * This component is on hold until an endpoint is implemented at
 * source-academy/cadet which is able to download and relay news and material
 * from lumiNUS to us.
 */
import { Card, NonIdealState, Spinner, Text } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import * as React from 'react';
import ContentDisplay, { IContentDisplayProps } from './commons/ContentDisplay';

export type Announcement = {
  author: string;
  title: string;
  content: string;
  pinned: boolean;
};

export interface IAnnouncementsProps {
  announcements?: Announcement[];
  handleAnnouncementsFetch: () => void;
}

interface IAnnouncementCardProps {
  announcements?: Announcement[];
}

class Announcements extends React.Component<IAnnouncementsProps, {}> {
  public render() {
    const props: IContentDisplayProps = {
      display: <AnnouncementCard announcements={this.props.announcements} />,
      loadContentDispatch: this.props.handleAnnouncementsFetch
    };
    return <ContentDisplay {...props} />;
  }
}

export const AnnouncementCard: React.SFC<IAnnouncementCardProps> = props => {
  if (props.announcements === undefined) {
    return <NonIdealState description="Fetching announcements..." visual={<Spinner />} />;
  } else if (props.announcements.length === 0) {
    return <NonIdealState title="There are no announcements." visual={IconNames.FEED} />;
  } else {
    const cards = props.announcements.map((ann, index) => (
      <div key={index}>
        <Card>
          <h4>{ann.title}</h4>
          <Text>{ann.content}</Text>
        </Card>
      </div>
    ));
    return <>{cards}</>;
  }
};

export default Announcements;
