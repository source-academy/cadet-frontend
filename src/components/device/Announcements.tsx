import { Card, NonIdealState, Spinner, Text } from '@blueprintjs/core'
import * as React from 'react'
import ContentDisplay, { IContentDisplayProps } from './ContentDisplay'

export type Announcement = {
  author: string
  title: string
  content: string
  pinned: boolean
}

export interface IAnnouncementsProps {
  announcements?: Announcement[]
  handleAnnouncementsFetch: () => void
}

interface IAnnouncementCardProps {
  announcements?: Announcement[]
}

class Announcements extends React.Component<IAnnouncementsProps, {}> {
  public render() {
    const props: IContentDisplayProps = {
      display: <AnnouncementCard announcements={this.props.announcements} />,
      loadContentDispatch: this.props.handleAnnouncementsFetch
    }
    return <ContentDisplay {...props} />
  }
}

export const AnnouncementCard: React.SFC<IAnnouncementCardProps> = props => {
  if (props.announcements === undefined) {
    return <NonIdealState description="Fetching announcements..." visual={<Spinner />} />
  } else if (props.announcements.length === 0) {
    return <NonIdealState title="There are no announcements." visual="feed" />
  } else {
    const cards = props.announcements.map((ann, index) => (
      <div key={index}>
        <Card>
          <h4>{ann.title}</h4>
          <Text>{ann.content}</Text>
        </Card>
      </div>
    ))
    return <>{cards}</>
  }
}

export default Announcements
