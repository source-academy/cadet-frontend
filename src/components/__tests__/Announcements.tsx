import * as React from 'react';

import { mount } from 'enzyme';

import Announcements, { IAnnouncementsProps } from '../Announcements';

const mockUndefinedAnnouncements: IAnnouncementsProps = {
  handleAnnouncementsFetch: () => {}
};

const mockEmptyAnnouncements: IAnnouncementsProps = {
  announcements: [],
  handleAnnouncementsFetch: () => {}
};

const mockPresentAnnouncements: IAnnouncementsProps = {
  announcements: [
    {
      author: 'Aministrator',
      title: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit',
      content:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam non tellus non ligula consectetur feugiat a at mauris. Ut sagittis, urna id suscipit dictum, ipsum diam sollicitudin enim, sed ultricies diam turpis vel libero. Maecenas finibus nisl ac lobortis ultricies. Integer molestie urna vitae nisi pharetra porttitor. Vestibulum massa diam, tristique quis ante a, posuere placerat magna. Pellentesque at lobortis purus, vitae imperdiet diam. Nulla eu rutrum neque. Aliquam efficitur consectetur ullamcorper. Pellentesque ultricies, diam ut vestibulum pellentesque, metus arcu laoreet ex, at mattis quam est non neque. Nam velit ipsum, posuere non porttitor commodo, lobortis in urna.Nulla facilisi. Donec mollis id nibh a luctus. Mauris vitae orci id velit pulvinar pellentesque non ut sapien. Curabitur eu consequat lorem. Proin pretium blandit scelerisque. Morbi ultricies, tellus non posuere pretium, felis sem convallis magna, ut sagittis elit felis et sem. Aliquam auctor suscipit condimentum. Nam posuere nulla dolor, in maximus risus feugiat vel. Phasellus vestibulum odio nec leo vehicula condimentum. Ut et semper libero. Cras fermentum mauris quis ex sodales, sit amet hendrerit augue lobortis. Maecenas eu dapibus enim, nec auctor est. Quisque quis erat erat. Curabitur sed rutrum felis, non venenatis diam. Fusce maximus rhoncus neque, in maximus velit semper eu. Fusce tempus lorem ut sodales pulvinar.',
      pinned: true
    },
    {
      author: 'Bob the builder',
      title: 'Can we fix it? A study forged by inquiry.',
      content:
        "Bob the Builder is a British children's animated television show created by Keith Chapman. In the original series, Bob appears in a stop motion animated programme as a building contractor, specialising in masonry, along with his colleague Wendy, various neighbours and friends, and their gang of anthropomorphised work-vehicles and equipment. The show is broadcast in many countries, but originates from the United Kingdom where Bob is voiced by English actor Neil Morrissey. The show was later created using CGI animation starting with the spin-off series Ready, Steady, Build!.",
      pinned: false
    }
  ],
  handleAnnouncementsFetch: () => {}
};

test('Announcements page "loading" content renders correctly', () => {
  const app = <Announcements {...mockUndefinedAnnouncements} />;
  const tree = mount(app);
  expect(tree.debug()).toMatchSnapshot();
});

test('Announcements page with 0 announcements renders correctly', () => {
  const app = <Announcements {...mockEmptyAnnouncements} />;
  const tree = mount(app);
  expect(tree.debug()).toMatchSnapshot();
});

test('Announcements page with multiple loaded announcements renders correctly', () => {
  const app = <Announcements {...mockPresentAnnouncements} />;
  const tree = mount(app);
  expect(tree.debug()).toMatchSnapshot();
});
