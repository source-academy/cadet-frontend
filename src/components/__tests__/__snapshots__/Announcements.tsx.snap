// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Announcements page "loading" content renders correctly 1`] = `
"<Announcements handleAnnouncementsFetch={[Function: handleAnnouncementsFetch]}>
  <ContentDisplay display={{...}} loadContentDispatch={[Function: handleAnnouncementsFetch]}>
    <div className=\\"ContentDisplay row center-xs\\">
      <div className=\\"col-xs-10 contentdisplay-content-parent\\">
        <Blueprint3.Card className=\\"contentdisplay-content\\" elevation={3} interactive={false}>
          <div className=\\"bp3-card bp3-elevation-3 contentdisplay-content\\">
            <Component announcements={[undefined]}>
              <Blueprint3.NonIdealState description=\\"Fetching announcements...\\" icon={{...}}>
                <div className=\\"bp3-non-ideal-state\\">
                  <div className=\\"bp3-non-ideal-state-visual\\">
                    <Blueprint3.Icon icon={{...}} iconSize={60}>
                      <Blueprint3.Spinner>
                        <div className=\\"bp3-spinner\\">
                          <div className=\\"bp3-spinner-animation\\">
                            <svg width={50} height={50} strokeWidth=\\"8.00\\" viewBox=\\"1.00 1.00 98.00 98.00\\">
                              <path className=\\"bp3-spinner-track\\" d=\\"M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90\\" />
                              <path className=\\"bp3-spinner-head\\" d=\\"M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90\\" pathLength={280} strokeDasharray=\\"280 280\\" strokeDashoffset={210} />
                            </svg>
                          </div>
                        </div>
                      </Blueprint3.Spinner>
                    </Blueprint3.Icon>
                  </div>
                  <div>
                    Fetching announcements...
                  </div>
                </div>
              </Blueprint3.NonIdealState>
            </Component>
          </div>
        </Blueprint3.Card>
      </div>
    </div>
  </ContentDisplay>
</Announcements>"
`;

exports[`Announcements page with 0 announcements renders correctly 1`] = `
"<Announcements announcements={{...}} handleAnnouncementsFetch={[Function: handleAnnouncementsFetch]}>
  <ContentDisplay display={{...}} loadContentDispatch={[Function: handleAnnouncementsFetch]}>
    <div className=\\"ContentDisplay row center-xs\\">
      <div className=\\"col-xs-10 contentdisplay-content-parent\\">
        <Blueprint3.Card className=\\"contentdisplay-content\\" elevation={3} interactive={false}>
          <div className=\\"bp3-card bp3-elevation-3 contentdisplay-content\\">
            <Component announcements={{...}}>
              <Blueprint3.NonIdealState title=\\"There are no announcements.\\" icon=\\"feed\\">
                <div className=\\"bp3-non-ideal-state\\">
                  <div className=\\"bp3-non-ideal-state-visual\\">
                    <Blueprint3.Icon icon=\\"feed\\" iconSize={60}>
                      <span icon=\\"feed\\" className=\\"bp3-icon bp3-icon-feed\\" title={[undefined]}>
                        <svg fill={[undefined]} data-icon=\\"feed\\" width={60} height={60} viewBox=\\"0 0 20 20\\">
                          <desc>
                            feed
                          </desc>
                          <path d=\\"M2.5 15a2.5 2.5 0 000 5 2.5 2.5 0 000-5zm.5-5c-.55 0-1 .45-1 1s.45 1 1 1c2.76 0 5 2.24 5 5 0 .55.45 1 1 1s1-.45 1-1c0-3.87-3.13-7-7-7zM3 0c-.55 0-1 .45-1 1s.45 1 1 1c8.28 0 15 6.72 15 15 0 .55.45 1 1 1s1-.45 1-1C20 7.61 12.39 0 3 0zm0 5c-.55 0-1 .45-1 1s.45 1 1 1c5.52 0 10 4.48 10 10 0 .55.45 1 1 1s1-.45 1-1C15 10.37 9.63 5 3 5z\\" fillRule=\\"evenodd\\" />
                        </svg>
                      </span>
                    </Blueprint3.Icon>
                  </div>
                  <Component>
                    <h4 className=\\"bp3-heading\\">
                      There are no announcements.
                    </h4>
                  </Component>
                </div>
              </Blueprint3.NonIdealState>
            </Component>
          </div>
        </Blueprint3.Card>
      </div>
    </div>
  </ContentDisplay>
</Announcements>"
`;

exports[`Announcements page with multiple loaded announcements renders correctly 1`] = `
"<Announcements announcements={{...}} handleAnnouncementsFetch={[Function: handleAnnouncementsFetch]}>
  <ContentDisplay display={{...}} loadContentDispatch={[Function: handleAnnouncementsFetch]}>
    <div className=\\"ContentDisplay row center-xs\\">
      <div className=\\"col-xs-10 contentdisplay-content-parent\\">
        <Blueprint3.Card className=\\"contentdisplay-content\\" elevation={3} interactive={false}>
          <div className=\\"bp3-card bp3-elevation-3 contentdisplay-content\\">
            <Component announcements={{...}}>
              <div>
                <Blueprint3.Card elevation={0} interactive={false}>
                  <div className=\\"bp3-card bp3-elevation-0\\">
                    <Component>
                      <h4 className=\\"bp3-heading\\">
                        Lorem ipsum dolor sit amet, consectetur adipiscing elit
                      </h4>
                    </Component>
                    <Blueprint3.Text>
                      <div className=\\"\\" title={[undefined]}>
                        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam non tellus non ligula consectetur feugiat a at mauris. Ut sagittis, urna id suscipit dictum, ipsum diam sollicitudin enim, sed ultricies diam turpis vel libero. Maecenas finibus nisl ac lobortis ultricies. Integer molestie urna vitae nisi pharetra porttitor. Vestibulum massa diam, tristique quis ante a, posuere placerat magna. Pellentesque at lobortis purus, vitae imperdiet diam. Nulla eu rutrum neque. Aliquam efficitur consectetur ullamcorper. Pellentesque ultricies, diam ut vestibulum pellentesque, metus arcu laoreet ex, at mattis quam est non neque. Nam velit ipsum, posuere non porttitor commodo, lobortis in urna.Nulla facilisi. Donec mollis id nibh a luctus. Mauris vitae orci id velit pulvinar pellentesque non ut sapien. Curabitur eu consequat lorem. Proin pretium blandit scelerisque. Morbi ultricies, tellus non posuere pretium, felis sem convallis magna, ut sagittis elit felis et sem. Aliquam auctor suscipit condimentum. Nam posuere nulla dolor, in maximus risus feugiat vel. Phasellus vestibulum odio nec leo vehicula condimentum. Ut et semper libero. Cras fermentum mauris quis ex sodales, sit amet hendrerit augue lobortis. Maecenas eu dapibus enim, nec auctor est. Quisque quis erat erat. Curabitur sed rutrum felis, non venenatis diam. Fusce maximus rhoncus neque, in maximus velit semper eu. Fusce tempus lorem ut sodales pulvinar.
                      </div>
                    </Blueprint3.Text>
                  </div>
                </Blueprint3.Card>
              </div>
              <div>
                <Blueprint3.Card elevation={0} interactive={false}>
                  <div className=\\"bp3-card bp3-elevation-0\\">
                    <Component>
                      <h4 className=\\"bp3-heading\\">
                        Can we fix it? A study forged by inquiry.
                      </h4>
                    </Component>
                    <Blueprint3.Text>
                      <div className=\\"\\" title={[undefined]}>
                        Bob the Builder is a British children&#39;s animated television show created by Keith Chapman. In the original series, Bob appears in a stop motion animated programme as a building contractor, specialising in masonry, along with his colleague Wendy, various neighbours and friends, and their gang of anthropomorphised work-vehicles and equipment. The show is broadcast in many countries, but originates from the United Kingdom where Bob is voiced by English actor Neil Morrissey. The show was later created using CGI animation starting with the spin-off series Ready, Steady, Build!.
                      </div>
                    </Blueprint3.Text>
                  </div>
                </Blueprint3.Card>
              </div>
            </Component>
          </div>
        </Blueprint3.Card>
      </div>
    </div>
  </ContentDisplay>
</Announcements>"
`;
